### PR TITLE
Allow jinja2 templates to be located in current conda environment

### PR DESCRIPTION
**Pull request by ukoethe**
Use case: Suppose one wants to build a set of packages with exactly identical toolset settings (same compiler options, tool versions etc.). This is easy to achieve by performing the build in a suitably prepared conda environment. However, conda does not yet provide a general method to query information about the current environment within `meta.yaml`. The present PR adds this capability. For example, let's say that installation of the toolset package places a template `toolset-info.yaml` in the environment's root directory. This template can now be loaded in `meta.yaml` via
```yaml
{% import 'CONDA_ENVIRONMENT/toolset-info.yaml' as toolset %}
```
and later be accessed via `{{ toolset.name }}`.

[At present, templates must be located in either `<conda_root>/Lib/site-packages/conda_build/templates` or `RECIPE_DIR`, which won't serve the above use case.]

### CODE REVIEW

**Comment by ukoethe** [comment]
I notice that the travis tests fail. Apparently, this is unrelated to the pull request, see issue #576.

File: conda_build/metadata.py [hunk 1]
```diff
@@ -269,9 +269,21 @@ def get_contents(meta_path):
...
     path, filename = os.path.split(meta_path)
-    loaders = [jinja2.PackageLoader('conda_build'),
+    loaders = [# search relative to '<conda_root>/Lib/site-packages/conda_build/templates'
+               jinja2.PackageLoader('conda_build'),
+               # search relative to RECIPE_DIR
                jinja2.FileSystemLoader(path)
                ]
+    
+    # search relative to current conda environment directory
+    conda_env_path = os.environ.get('CONDA_ENV_PATH',  # path to current conda environment
```
Comment (asmeurer) [inline conda_build/metadata.py]: Should this just be `PREFIX`. Which conda environment is this referring to?

Comment (ukoethe) [inline conda_build/metadata.py]: When I call `activate foo`, conda automatically defines `CONDA_ENV_PATH=<conda_root>/envs/foo`. Is this non-portable?

Using the `PREFIX` doesn't work because the `_build` directory is not yet populated when `meta.yaml` is parsed, so one can't find any file there. Python's `sys.prefix` is unsuitable as well because the template mechanism should also work for non-python builds (such as `libtiff`).

Comment (asmeurer) [inline conda_build/metadata.py]: So is this supposed to be `_build` or whatever conda environment happens to be active when you run the build?

Comment (ukoethe) [inline conda_build/metadata.py]: It refers to the active environment where I run the build. 

Background: On Windows, binaries from different compiler versions tend to be incompatible. For example, when I use the official `libtiff` or `libpng` (created with Visual Studio 2008) in a build with Visual Studio 2012, the resulting program crashes. In order to support multiple compilers, I want to create environments like `vc9` (for Visual Studio 2008) and `vc11` (for Visual Studio 2012). Before each build, I activate the appropriate environment. Then `CONDA_ENVIRONMENT/toolset-info.yaml` contains all the necessary information to guarantee consistent builds.

Comment (ukoethe) [inline conda_build/metadata.py]: FWIW: I know that Python 2.7 doesn't officially support Visual Studio 2010 or newer. However, this is easily fixed. We need more modern compilers because some extension modules can't be compiled with VS 2008.

File: conda_build/metadata.py [hunk 2]
```diff
@@ -269,9 +269,21 @@ def get_contents(meta_path):
...
+                     os.environ.get('CONDA_PATH',      # else: path to conda root environment
```
Comment (ukoethe) [inline conda_build/metadata.py]: It seems that `CONDA_PATH` is not necessarily defined. Unless there is another suitable fallback, this line should be dropped.

Comment (asmeurer) [inline conda_build/metadata.py]: I agree.

**Comment by asmeurer** [comment]
This is not how I would solve this problem, but since you are going all in with the jinja2 templating and this makes things easier for you, it seems fine to me.

File: conda_build/build.py [hunk 3]
```diff
@@ -500,6 +501,8 @@ def test(m, verbose=True, channel_urls=(), override_channels=False):
 
     tmp_dir = join(config.croot, 'test-tmp_dir')
     rm_rf(tmp_dir)
+    time.sleep(2)  # give the OS some time to complete the rm_rf() command
```
Comment (asmeurer) [inline conda_build/build.py]: I don't like this fix, though. Can you split this into a separate PR?

Comment (ukoethe) [inline conda_build/build.py]: Sure. How would you solve the problem? It is a showstopper for fully automated builds.

Comment (asmeurer) [inline conda_build/build.py]: Is there an issue with more information on what the problem is? If anything I'd like to fix this in rm_rf (ideally by doing a check, rather than sleeping by an arbitrary amount of time).

**Comment by ukoethe** [comment]
I'm open for better suggestions than the toolset-info approach. 

I already experimented with conda's `features` capability, but all too often `conda install` selected the wrong build. Just to make sure that I understand `features` correctly: suppose feature 'vc11' is tracked in the current environment and there are versions
`foo-1.2.0`,  `foo-1.2.0-vc11` (defines feature 'vc11'), and `foo-1.4.1`
Is it true that
`conda install foo`
will always select the highest version number (here: 1.4.1), regardless of the features currently being tracked? That is, if no package among those with highest version number defines feature 'vc11', conda won't downgrade the version number, but would instead select a package without that feature? (Sorry if these questions are stupid, but documentation of the `features` capability is almost non-existent.)

**Comment by asmeurer** [comment]
No, if there is a feature for a package available and that feature is enabled, conda will only look at the versions of the package with that feature.

File: conda_build/build.py (hunk 3 above)
Comment (ukoethe) [inline conda_build/build.py]: Here is the error message:
```
BUILD END: use_foo-1.0-0
An unexpected error has occurred, please consider sending the
following traceback to the conda GitHub issue tracker at:

    https://github.com/conda/conda-build/issues

Include the output of the command 'conda info' in your report.


Traceback (most recent call last):
  File "C:\conda\Scripts\conda-build-script.py", line 4, in <module>
    sys.exit(main())
  File "C:\conda\lib\site-packages\conda_build\main_build.py", line 190, in main
    args_func(args, p)
  File "C:\conda\lib\site-packages\conda_build\main_build.py", line 432, in args_func
    args.func(args, p)
  File "C:\conda\lib\site-packages\conda_build\main_build.py", line 419, in execute
    channel_urls=channel_urls, override_channels=args.override_channels)
  File "C:\conda\lib\site-packages\conda_build\build.py", line 506, in test
    os.makedirs(tmp_dir)
  File "C:\conda\lib\os.py", line 157, in makedirs
    mkdir(name, mode)
WindowsError: [Error 5] Access denied: 'C:\\conda\\conda-bld\\test-tmp_dir'

> conda info
Current conda install:

             platform : win-64
        conda version : 3.16.0-36-gd28d3f3-dirty
  conda-build version : 1.17.0+49.g0495f64
       python version : 2.7.9.final.0
     requests version : 2.7.0
     root environment : C:\conda  (writable)
  default environment : C:\conda\envs\foo
     envs directories : C:\conda\envs
        package cache : C:\conda\pkgs
         channel URLs : https://repo.continuum.io/pkgs/free/win-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/win-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : None
    is foreign system : False
```
The error only occurs with `test-tmp_dir`. Moreover, it seems that it only happens when the package doesn't define tests, but I'm not sure.

Comment (ukoethe) [inline conda_build/build.py]: see #577

**Comment by ukoethe** [comment]
According to my experiments, this is not the case (I just repeated them to make sure I'm not missing something. I have build packages 
```
track_vc11-1.0-0.tar.bz2
foo-1.0-0.tar.bz2
foo-1.0-vc11.tar.bz2
foo-2.0-0.tar.bz2
use_foo-1.0-vc11.tar.bz2
```
created from the following `meta.yaml` files:
```yaml
package:
  name:  track_vc11
  version: "1.0"
build:
  features:
    - vc11
  track_features:
    - vc11
```
```yaml
package:
  name:  foo
  version: "1.0"
```
```yaml
package:
  name:  foo
  version: "1.0"
build:
  string: vc11
  features:
    - vc11
requirements:
  build:
    - track_vc11
  run:
    - track_vc11
```
```yaml
package:
  name:  foo
  version: "2.0"
```
```yaml
package:
  name:  use_foo
  version: "1.0"
build:
  string: vc11
  features:
    - vc11
requirements:
  build:
    - track_vc11
    - foo
  run:
    - track_vc11
    - foo
```
Now I create an environment where `vc11` is being tracked:
```
conda create --use-local -n vc11 track_vc11
activate vc11
```
Within this environment, I install `use_foo` and get:
```
The following NEW packages will be INSTALLED:

    foo:     2.0-0
    use_foo: 1.0-vc11 [vc11]
```
According to your explanation above, I'd expect to see `foo:  1.0-vc11 [vc11]`.

**Comment by ukoethe** [comment]
I suspect this is a conda bug. See issue https://github.com/conda/conda/issues/1589.

**Comment by stuarteberg** [comment]
@ukoethe, Now that #578 has been merged, can this PR (#575) be closed?

**Comment by ukoethe** [comment]
Indeed, thanks for the reminder.
